### PR TITLE
atomic_helper.py: ensure presence of parent directories

### PR DESCRIPTION
**Pull request by sshedi**
## Proposed Commit Message
```
fix(atomic_helper.py): ensure presence of parent directories
```

## Test Steps
```
Steps to reproduce:

$ cloud-init clean -ls
$ cloud-init -d modules -m init
2024-02-22 18:59:46,522 - handlers.py[DEBUG]: start: modules-init: running modules for init
2024-02-22 18:59:46,522 - util.py[DEBUG]: Reading from /proc/uptime (quiet=False)
....
[('set_hostname', FileNotFoundError(2, 'No such file or directory'))]
```

## Checklist

- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [x] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Review by aciba90 — COMMENTED** [review]
`cloud-init modules` is expected to run after `cloud-init init --local` and `cloud-init init` and be driven by init-system services.

The config modules corresponding to the init phase, `cloud-init modules -m init`, are run as part of `cloud-init init`.

Not sure if we want to support this path in general, if yes, I have some concerns as:

1. What's the performance impact of this change.
2. What about the permissions of created parent directories? 
3. Is this the correct place to do this, or would be better to ensure that necessary parent directories are pre-created upper in the call stack? or should we control the creation of the parent dir with a parameter in this function.

That said, why not just run `cloud-init init --local`, which will create the paths you are missing, before executing `cloud-init modules -m init`.

**Comment by sshedi** [comment]
Thanks for the inputs and clarifications. Please find my replies inline.

> `cloud-init modules` is expected to run after `cloud-init init --local` and `cloud-init init` and be driven by init-system services.
> 
[sshedi]: In that case, `cloud-init init` should not be allowed to run before running `cloud-init --local` and it should be validated, right? Please correct me if my understanding is wrong.

> The config modules corresponding to the init phase, `cloud-init modules -m init`, are run as part of `cloud-init init`.
> 
> Not sure if we want to support this path in general, if yes, I have some concerns as:
> 
[sshedi]: Even if it's not supported, this should not end in a traceback. There is possibility of a use case where someone might use cloud-init for setting hostname only and nothing else.

>     1. What's the performance impact of this change.
> 
[sshedi]: should not be much, if it's done in the order you mentioned, this directory gets created when `local` stage is run but now this is happening here.

>     2. What about the permissions of created parent directories?
> 
[sshedi]: this is creating the missing `data` dir under `/var/lib/cloud/` and with and without this change, the permissions are 755.
>     3. Is this the correct place to do this, or would be better to ensure that necessary parent directories are pre-created upper in the call stack? or should we control the creation of the parent dir with a parameter in this function.
> 

[sshedi]: Adding a parameter to this function might be an overkill, enure_dir will create the dir only if the dir doesn't exist.
My question here is, should we use `ensure_dirs` instead.

> 
> That said, why not just run `cloud-init init --local`, which will create the paths you are missing, before executing `cloud-init modules -m init`.
[sshedi]: Like I mentioned above, it's a debatable topic. Running `--local` sounds reasonable but it'll be good if set_hostname without a traceback.

**Review by aciba90 — CHANGES_REQUESTED** [review]
Thanks, @sshedi!

> cloud-init modules is expected to run after cloud-init init --local and cloud-init init and be driven by init-system services.

>> [sshedi]: In that case, cloud-init init should not be allowed to run before running cloud-init --local and it should be validated, right? Please correct me if my understanding is wrong.

A normal execution path is:

1. cloud-init init --local
2. cloud-init init
3. cloud-init modules --mode config
4. cloud-init modules --mode final

in that order, and the init config modules are executed as part of `cloud-init init`. `cloud-init modules --mode init` is a command mostly for debugging purposes that we are going to [deprecate](https://github.com/canonical/cloud-init/commit/0c80569929c8b40d244c49af2aac75c6928f06ab) to avoid confusion.

See [boot-stages](https://docs.cloud-init.io/en/latest/explanation/boot.html) for more info.

> [sshedi]: Even if it's not supported, this should not end in a traceback. There is possibility of a use case where someone might use cloud-init for setting hostname only and nothing else.

I agree, we shouldn't end up with a traceback. Maybe `cloud-init single --name set_hostname --frequency always` is more appropiate to your use case, if you only want to set the hostname, see [single](https://docs.cloud-init.io/en/latest/reference/cli.html#single). But, if `cloud-init init --local` / `cloud-init init` did not run, some ds meta-data / cloud-config could be not yet fetched.

> 2. What about the permissions of created parent directories?

>> [sshedi]: this is creating the missing data dir under /var/lib/cloud/ and with and without this change, the permissions are 755.

The concern was about other code paths creating folders on other locations with opener permissions, I have double-checked that the main folders that cloud-init interacts are all with `755 root:root`, so we are safe with `util.ensure_dir`.

```
[drwxr-xr-x root     root    ]  /etc/cloud/
├── [drwxr-xr-x root     root    ]  clean.d
├── [drwxr-xr-x root     root    ]  cloud.cfg.d
└── [drwxr-xr-x root     root    ]  templates
[drwxr-xr-x root     root    ]  /var/run/cloud-init/
└── [drwxr-xr-x root     root    ]  sem
[drwxr-xr-x root     root    ]  /var/lib/cloud/
├── [drwxr-xr-x root     root    ]  data
├── [drwxr-xr-x root     root    ]  handlers
├── [lrwxrwxrwx root     root    ]  instance -> /var/lib/cloud/instances/5b18e235-17ad-41b5-a5b7-44569c20bdf3
├── [drwxr-xr-x root     root    ]  instances
│   └── [drwxr-xr-x root     root    ]  5b18e235-17ad-41b5-a5b7-44569c20bdf3
│       ├── [drwxr-xr-x root     root    ]  handlers
│       ├── [drwxr-xr-x root     root    ]  scripts
│       └── [drwxr-xr-x root     root    ]  sem
├── [drwxr-xr-x root     root    ]  scripts
│   ├── [drwxr-xr-x root     root    ]  per-boot
│   ├── [drwxr-xr-x root     root    ]  per-instance
│   ├── [drwxr-xr-x root     root    ]  per-once
│   └── [drwxr-xr-x root     root    ]  vendor
├── [drwxr-xr-x root     root    ]  seed
└── [drwxr-xr-x root     root    ]  sem
```

> [sshedi]: Adding a parameter to this function might be an overkill, enure_dir will create the dir only if the dir doesn't exist.
My question here is, should we use ensure_dirs instead.

Yeah, I think `util.ensure_dir` is more appropriate as we know we are going to create only one directory.

File: cloudinit/atomic_helper.py [hunk 1]
```diff
@@ -43,9 +45,9 @@ def write_file(
 
     tf = None
     try:
-        tf = tempfile.NamedTemporaryFile(
-            dir=os.path.dirname(filename), delete=False, mode=omode
-        )
+        dirname = os.path.dirname(filename)
+        util.ensure_dirs([dirname])
```
[the remark this answers is not in the document]
Comment (sshedi) [reply cloudinit/atomic_helper.py]: Thanks for the suggestions and recommendations. I have made the required code changes.

**Review by aciba90 — APPROVED** [review]
LGTM, thanks @sshedi!

**Comment by sshedi** [comment]
Thanks for taking the fix 🙏 Have a nice time ahead.

**Pull request merged**